### ISSUE TITLE
Support 0d tensors in reductionOp

### ISSUE THREAD

**Issue by IvanYashchuk** [csarofeen/pytorch#1768]
### 🚀 The feature, motivation and pitch

PyTorch allows calling reduction on 0d tensors:
```py
In [1]: import torch

In [2]: a = torch.randn(())

In [3]: a.ndim
Out[3]: 0

In [4]: torch.sum(a, 0, keepdim=True)
Out[4]: tensor(0.5233)

In [5]: torch.sum(a, 0, keepdim=False)
Out[5]: tensor(0.5233)
```

But nvFuser raises an error `"Tried to reduce a 0-dim tensor"`.

Here's a Python reproducer:
```py
import torch
from torch._C._nvfuser import Fusion, FusionDefinition, DataType

# Construct and Define Fusion
fusion = Fusion()

with FusionDefinition(fusion) as fd :
    t0 = fd.define_tensor(0)
    fd.add_input(t0)
    t1 = fd.Ops.sum(t0, [0], False, DataType.Float)
    fd.add_output(t1)

fusion.print_ir()

# Execute Fusion
input1 = torch.ones((), device='cuda')
outputs = fusion.execute([input1, ])

print(outputs)
```

Using the following patch, that disables ndim checks, makes my tests pass and the above Python example works. Is it safe to relax the requirements on the dimensionality for reduction?

```diff
index 3bab78b566..a6fc269d38 100644
--- a/torch/csrc/jit/codegen/cuda/arith.cpp
+++ b/torch/csrc/jit/codegen/cuda/arith.cpp
@@ -867,7 +867,7 @@ static TensorView* newForReduction(
       "Asked for ouput of reduction, but no reduction axis provided.");

   TORCH_INTERNAL_ASSERT(
-      (*(axes_set.rbegin())) < orig_domain.size(),
+      (*(axes_set.rbegin())) <= orig_domain.size(),
       "Error setting up reduction, reduction axis (",
       *(axes_set.rbegin()),
       ") is outside nDims (",
@@ -921,7 +921,7 @@ TensorView* reductionOp(
       TensorDomain::sameAs(tv->getMaybeRFactorDomain(), tv->domain()->domain()),
       "Reducing a tensor once it's gone under transformations is not permitted at this time. Please set reductions before calling split/merge/computeAt.");

-  TORCH_CHECK(tv->nDims() > 0, "Tried to reduce a 0-dim tensor");
+  // TORCH_CHECK(tv->nDims() > 0, "Tried to reduce a 0-dim tensor");

   TORCH_CHECK(axes.size() > 0, "No reduction axis specified");

@@ -933,7 +933,7 @@ TensorView* reductionOp(
     }

     TORCH_CHECK(
-        axis >= 0 && axis < ndims,
+        axis >= 0 && axis <= ndims,
         "Reduction on invalid axis, recieved: ",
         axis,
         " however tensor view only has ",
```

**Comment by naoyam** [comment]
Hmm, what's the exact semantic of reducing 0-dim tensors? Obviously, there's no dimension, so nothing can be reduced. When you do:

```
In [4]: torch.sum(a, 0, keepdim=True)
Out[4]: tensor(0.5233)
```

I assume `0` here is supposed to specify the dimension to reduce, but is it just ignored? I tried `torch.sum(a, 1)`, but it resulted in an error, which seems to indicate `0` is handled specially.

**Comment by IvanYashchuk** [comment]
It seems the convention is that the `dim` argument for 0d tensors can be either `-1` or `0`. I can't find the documentation about this. `keepdim` also doesn't affect the shape in this case, because `keepdim=True` means that the shape doesn't change, and `keepdim=False` means that the shape is reduced but 0d is the smallest dimensional tensor.

The reduction of a 0-dim tensor using some binary op and initial (neutral) element should be thought of as just an application of that binary op with the value of a tensor and initial element once: for the sum, the binary op is "+" operation and the initial element is 0, `+(0, a) -> a`.

**Comment by naoyam** [comment]
Thanks. So, we should be able to just ignore the operation as long as `dim` is `-1` or `0`. This seems like something that should be handled at the integration. What do you think, @jjsjann123?

**Comment by jjsjann123** [comment]
> ```
> with FusionDefinition(fusion) as fd :
>    t0 = fd.define_tensor(0)
>    fd.add_input(t0)
>    t1 = fd.Ops.sum(t0, [0], False, DataType.Float)
>    fd.add_output(t1)
> ```

we would want to allow user to do this. If we are exposing APIs at python level, it's tricky to hide it behind something else then.

If we are worried about scheduler, maybe one thing we can try it to update sum in `arith.cpp` and make it a copy node `set` instead. Thoughts ? @kevinstephano
